File: webui/src/main/scala/com/taskadapter/webui/pages/SchedulesListPage.scala
package com.taskadapter.webui.pages

import com.taskadapter.vaadin14shim.Label
import com.taskadapter.vaadin14shim.VerticalLayout
import com.taskadapter.vaadin14shim.HorizontalLayout
import java.util.UUID

import com.taskadapter.web.uiapi.{ConfigId, Schedule}
import com.taskadapter.webui._
import com.taskadapter.webui.service.Preservices
import com.vaadin.data.sort.SortOrder
import com.vaadin.data.util.{BeanItem, BeanItemContainer}
import com.vaadin.shared.data.sort.SortDirection
import com.vaadin.ui._
import org.slf4j.LoggerFactory

import scala.beans.BeanProperty
import scala.collection.JavaConverters._

case class ScheduleListItem(id: String, configId: ConfigId,
                            @BeanProperty configLabel: String,
                            @BeanProperty intervalMin: Int,
                            @BeanProperty to: String)

class SchedulesListPage() extends BasePage {
  private val configOps: ConfigOperations = SessionController.buildConfigOperations()
  private val services: Preservices = SessionController.getServices

  EventTracker.trackPage("schedules");
  private val log = LoggerFactory.getLogger(classOf[SchedulesListPage])
  private val configRowsToShowInListSelect = 15

  val ds = new BeanItemContainer[ScheduleListItem](classOf[ScheduleListItem])
  val grid = new Grid(ds)

  val ui = new VerticalLayout()
  val listLayout = createListLayout()
  val configsListSelect = createConfigsList()

  ui.addComponent(listLayout)
  showSchedules(services.schedulesStorage.getSchedules())

  def showSchedules(results: Seq[Schedule]): Unit = {
    ds.removeAllItems()
    ds.addAll(
      results.flatMap { schedule =>
        val maybeConfig = configOps.getOwnedConfigs.find(c => c.configId == schedule.configId)
        if (maybeConfig.isDefined) {
          Some(ScheduleListItem(schedule.id, schedule.configId, maybeConfig.get.label,
            schedule.intervalInMinutes, maybeConfig.get.getConnector2.getLabel))
        } else {
          log.error(s"cannot find config for schedule $schedule")
          None
        }
      }.asJava)
    grid.setSortOrder(List(new SortOrder("configLabel", SortDirection.ASCENDING)).asJava)
    applyUI(listLayout)
  }


  def showSchedules(): Unit = {
    showSchedules(services.schedulesStorage.getSchedules())
  }

  private def applyUI(newUi: Component): Unit = {
    ui.removeAll()
    ui.add(newUi)
  }

  private def showSchedule(scheduleId: String): Unit = {
    EventTracker.trackPage("edit_schedule_page")
    val schedule = services.schedulesStorage.get(scheduleId).get
    showSchedule(schedule)
  }

  private def showNewScheduleEditor(configId: ConfigId): Unit = {
    EventTracker.trackPage("create_schedule_page")
    val schedule = Schedule(UUID.randomUUID().toString, configId, 60, false, false)
    showSchedule(schedule)
  }

  private def showSchedule(schedule: Schedule): Unit = {
    val config = configOps.getOwnedConfigs.find(c => c.configId == schedule.configId).get
    val editSchedulePage = new EditSchedulePage(
      config.label,
      config.getConnector1.getLabel,
      config.getConnector2.getLabel,
      schedule, (schedule) => saveSchedule(schedule),
      showSchedules,
      (schedule) => deleteSchedule(schedule)
    )
    applyUI(editSchedulePage.ui)
  }

  def saveSchedule(schedule: Schedule): Unit = {
    services.schedulesStorage.store(schedule)
    showSchedules()
  }

  def deleteSchedule(schedule: Schedule): Unit = {
    services.schedulesStorage.delete(schedule.id)
    showSchedules()
  }

  def getSelectedConfigId(): ConfigId = {
    configsListSelect.getValue.asInstanceOf[ConfigId]
  }

  def showSelectConfig(): Unit = {
    val layout = new VerticalLayout()
    layout.setSpacing(true)
    val goButton = new Button(Page.message("schedules.newConfig.go"))
    goButton.addClickListener(_ => showNewScheduleEditor(getSelectedConfigId()))

    val label = new Label(Page.message("schedules.selectConfig.title"))
    label.addClassName(Sizes.tabIntro)
    layout.add(label)
    val horizontalLayout = new HorizontalLayout(configsListSelect, goButton)
    horizontalLayout.setSpacing(true)
    layout.add(horizontalLayout)

    reloadConfigsInList()
    applyUI(layout)
  }

  private def createConfigsList(): ListSelect = {
    val res = new ListSelect()
    res.setNullSelectionAllowed(false)
    res
  }

  private def reloadConfigsInList(): Unit = {
    configOps.getOwnedConfigs.foreach { s =>
      configsListSelect.addItem(s.configId)
      configsListSelect.setItemCaption(s.configId, if (s.label.isEmpty) {
        Page.message("schedules.configsList.defaultLabelForConfigs")
      } else {
        s.label
      }
      )
    }
    val rowsNumber = if (configsListSelect.size < configRowsToShowInListSelect) {
      configsListSelect.size
    } else {
      configRowsToShowInListSelect
    }
    configsListSelect.setRows(rowsNumber)
    if (configsListSelect.getRows > 0) {
      configsListSelect.select(configsListSelect.getItemIds.iterator().next())
    }
  }

  def createListLayout(): Layout = {

    val addButton = new Button(Page.message("schedules.newButton"))
    addButton.addClickListener(_ => showSelectConfig())

    val checkbox = new CheckBox(Page.message("schedules.scheduledEnabled"))
    checkbox.setValue(services.settingsManager.schedulerEnabled)
    checkbox.setImmediate(true)
    checkbox.addValueChangeListener(_ => services.settingsManager.setSchedulerEnabled(checkbox.getValue))

    grid.setWidth("980px")
    grid.removeAllColumns()

    grid.addColumn("configLabel")
      .setHeaderCaption(Page.message("schedules.column.label"))
      .setExpandRatio(1)


    grid.addColumn("intervalMin")
      .setHeaderCaption(Page.message("schedules.column.interval"))
      .setExpandRatio(1)

    //    grid.addColumn("to")
    //      .setHeaderCaption(Page.message("schedules.column.to"))
    //      .setExpandRatio(2)

    grid.addSelectionListener { _ =>
      val row = grid.getSelectedRow
      if (row != null) {
        val result = grid.getContainerDataSource.getItem(row).asInstanceOf[BeanItem[ScheduleListItem]].getBean
        showSchedule(result.id)
      }
    }

    val label = new Label(Page.message("schedules.intro"))
    label.addClassName(Sizes.tabIntro)

    val controlsLayout = new HorizontalLayout(addButton, checkbox)
    controlsLayout.setWidth("100%")
    controlsLayout.setSpacing(true)
    controlsLayout.setComponentAlignment(addButton, Alignment.MIDDLE_LEFT)
    controlsLayout.setComponentAlignment(checkbox, Alignment.MIDDLE_RIGHT)

    val listLayout = new VerticalLayout()
    listLayout.add(label)
    listLayout.add(controlsLayout)
    listLayout.add(grid)
    listLayout
  }

}

File: webui/src/main/java/com/taskadapter/webui/pages/SchedulesListPage.java
package com.taskadapter.webui.pages;

import com.taskadapter.web.PopupDialog;
import com.taskadapter.web.uiapi.ConfigId;
import com.taskadapter.web.uiapi.Schedule;
import com.taskadapter.web.uiapi.UISyncConfig;
import com.taskadapter.webui.BasePage;
import com.taskadapter.webui.ConfigOperations;
import com.taskadapter.webui.Layout;
import com.taskadapter.webui.Page;
import com.taskadapter.webui.SessionController;
import com.taskadapter.webui.Sizes;
import com.taskadapter.webui.service.Preservices;
import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.checkbox.Checkbox;
import com.vaadin.flow.component.dependency.CssImport;
import com.vaadin.flow.component.grid.Grid;
import com.vaadin.flow.component.grid.GridSortOrder;
import com.vaadin.flow.component.html.Label;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;
import com.vaadin.flow.data.provider.SortDirection;
import com.vaadin.flow.data.selection.SelectionListener;
import com.vaadin.flow.router.Route;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

@Route(value = Navigator.SCHEDULES_LIST, layout = Layout.class)
@CssImport(value = "./styles/views/mytheme.css")
public class SchedulesListPage extends BasePage {
    private static Logger log = LoggerFactory.getLogger(SchedulesListPage.class);
    private static final int configRowsToShowInListSelect = 15;

    private ConfigOperations configOps = SessionController.buildConfigOperations();
    private Preservices services = SessionController.getServices();

    private Grid<ScheduleListItem> grid = new Grid<>();

    private Component listLayout;
    private VerticalLayout configsListLayout = new VerticalLayout();
    private Grid.Column<ScheduleListItem> configLabelColumn;

    public SchedulesListPage() {
        buildUi();
    }

    private void buildUi() {
        listLayout = createListLayout();
        showInPage(listLayout);
    }

    private void showInPage(Component component) {
        removeAll();
        add(LayoutsUtil.centered(Sizes.mainWidth(), component));
    }

    @Override
    protected void beforeEnter() {
        showSchedules(services.schedulesStorage.getSchedules());
    }

    private void saveSchedule(Schedule schedule) {
        services.schedulesStorage.store(schedule);
        showSchedules();
    }

    private void deleteSchedule(Schedule schedule) {
        services.schedulesStorage.delete(schedule.getId());
        showSchedules();
    }

    private void showSchedule(final Schedule schedule) {
        UISyncConfig config = getConfigForSchedule(schedule)
                .orElseThrow(() -> new RuntimeException("cannot open schedule " + schedule
                        + " because its config is not found"));
        EditSchedulePage editSchedulePage = new EditSchedulePage(
                config.label(),
                config.getConnector1().getLabel(),
                config.getConnector2().getLabel(),
                schedule,
                saveSchedule -> {
                    saveSchedule(saveSchedule);
                    return null;
                },
                () -> showSchedules(),
                deleteSchedule -> {
                    PopupDialog.confirm("Do you want to delete this scheduled item?",
                            () -> {
                                deleteSchedule(deleteSchedule);
                                return null;
                            });
                    return null;
                }
        );
        showInPage(editSchedulePage.ui());
    }

    private Optional<UISyncConfig> getConfigForSchedule(Schedule schedule) {
        return configOps.getOwnedConfigs()
                .stream().filter(c -> c.configId().equals(schedule.getConfigId()))
                .findFirst();
    }

    private void showSchedule(String scheduleId) {
        Schedule schedule = services.schedulesStorage.get(scheduleId).get();
        showSchedule(schedule);
    }

    private void showNewScheduleEditor(ConfigId configId) {
        Schedule schedule = new Schedule(UUID.randomUUID().toString(), configId, 60, false, false);
        showSchedule(schedule);
    }

    private void showSchedules() {
        showSchedules(services.schedulesStorage.getSchedules());
    }

    private void showSelectConfig() {
        VerticalLayout layout = new VerticalLayout();
        layout.setSpacing(true);

        Label label = new Label(Page.message("schedules.selectConfig.title"));
        layout.add(label);
        layout.add(configsListLayout);

        showConfigsInList();
        showInPage(layout);
    }

    private Component createListLayout() {

        Button addButton = new Button(Page.message("schedules.newButton"),
                event -> showSelectConfig());

        Checkbox checkbox = new Checkbox(Page.message("schedules.scheduledEnabled"));
        checkbox.setValue(services.settingsManager.schedulerEnabled());
        checkbox.addValueChangeListener(value ->
                services.settingsManager.setSchedulerEnabled(checkbox.getValue()));

        grid.setWidth("980px");
        configLabelColumn = grid.addColumn(ScheduleListItem::getConfigLabel)
                .setHeader(Page.message("schedules.column.label"))
                .setFlexGrow(1);

        grid.addColumn(ScheduleListItem::getIntervalMin)
                .setHeader(Page.message("schedules.column.interval"))
                .setFlexGrow(1);

        grid.addSelectionListener((SelectionListener<Grid<ScheduleListItem>, ScheduleListItem>) event ->
                event.getFirstSelectedItem()
                        .ifPresent(item -> showSchedule(item.id)));

        Label label = new Label(Page.message("schedules.intro"));

        HorizontalLayout controlsLayout = new HorizontalLayout(addButton, checkbox);
        controlsLayout.setWidth("100%");
        controlsLayout.setSpacing(true);

        VerticalLayout listLayout = new VerticalLayout();
        listLayout.add(label);
        listLayout.add(controlsLayout);
        listLayout.add(grid);
        return listLayout;
    }

    private void showConfigsInList() {
        List<UISyncConfig> configsList = configOps.getOwnedConfigs();

        configsList.stream().forEach(c -> {
            Button button = new Button(c.getLabel());
            button.setWidth("200px");
            button.addClickListener(e -> showNewScheduleEditor(c.configId()));
            configsListLayout.add(button);
        });
    }

    private void showSchedules(List<Schedule> results) {
        List<ScheduleListItem> items = results.stream().filter(schedule -> {
            Optional<UISyncConfig> maybeConfig = getConfigForSchedule(schedule);
            if (!maybeConfig.isPresent()) {
                log.error("cannot find config for schedule " + schedule);
            }
            return maybeConfig.isPresent();
        }).map(schedule -> {
            // TODO 14 do not load the configs twice!
            Optional<UISyncConfig> maybeConfig = getConfigForSchedule(schedule);
            return new ScheduleListItem(schedule.getId(), schedule.getConfigId(), maybeConfig.get().label(),
                    schedule.getIntervalInMinutes(), maybeConfig.get().getConnector2().getLabel());
        }).collect(Collectors.toList());

        grid.sort(Arrays.asList(new GridSortOrder(configLabelColumn, SortDirection.ASCENDING)));

        grid.setItems(items);
        showInPage(listLayout);
    }

    private static class ScheduleListItem {
        String id;
        ConfigId configId;
        String configLabel;
        int intervalMin;
        String to;

        public ScheduleListItem(String id, ConfigId configId, String configLabel, int intervalMin, String to) {
            this.id = id;
            this.configId = configId;
            this.configLabel = configLabel;
            this.intervalMin = intervalMin;
            this.to = to;
        }

        public String getId() {
            return id;
        }

        public ConfigId getConfigId() {
            return configId;
        }

        public String getConfigLabel() {
            return configLabel;
        }

        public int getIntervalMin() {
            return intervalMin;
        }

        public String getTo() {
            return to;
        }
    }
}

